portal/portal-charon/charon/src/webapp/scripts/themeswitcher.js

```diff
@@ -1,43 +1,97 @@
-function themeSwitcher(){
+function sakaiThemeSwitcher(){
 
-    let themeSwitcher = document.getElementById("themeSwitcher");
+    let darkThemeSwitcher = document.getElementById("sakai-darkThemeSwitcher");
 
-    SAK_THEME_SELECTION_KEY = portal.user.id + "#theme";
+    const defaultThemeClass = 'sakaiUserTheme-notSet';
+    const lightThemeClass = 'sakaiUserTheme-light';
+    const darkThemeClass = 'sakaiUserTheme-dark';
 
-    if(themeSwitcher != null) {
+    init();
+    
+    function init() {
+        // attach listener to dark theme toggle switch
+        darkThemeSwitcher && darkThemeSwitcher.addEventListener('click', toggleDarkTheme, false);
+
+        if(isLoggedIn()) {
+            // only check for unset theme preference because light and dark themes are already set by Java
+            if(isPortalThemeUserPrefUnset()) {
+                if(isOsDarkThemeSet()) {
+                    enableDarkTheme();
+                } else {
+                    // to define a user preference:
+                    setPortalThemeUserPref(lightThemeClass);
+                }
+            }
+        } else if(isOsDarkThemeSet()) {
+            addCssClassToMarkup(darkThemeClass);
+        }
         
-        initTheme(); // on page load, if user has already selected a specific theme -> apply it
+        if (document.documentElement.classList.contains(darkThemeClass)) {
+            // the dark theme switch toggle is off by default, so toggle to on if dark theme is enabled
+            setDarkThemeSwitcherToggle(true);
+        }
+    }
 
-        themeSwitcher.addEventListener('click', setTheme, false);
+    function addCssClassToMarkup(themeClass) {
+        document.documentElement.classList.add(themeClass);
+    }
 
+    function removeCssClassFromMarkup(themeClass) {
+        document.documentElement.classList.remove(themeClass);
     }
 
-    function setTheme() {
+    function isOsDarkThemeSet() {
+        return window.matchMedia('(prefers-color-scheme: dark)').matches;
+    }
+
+    function isLoggedIn() {
+        return portal.user.id;
+    }
 
-        // Set to dark mode
-        if (themeSwitcher.getAttribute("aria-checked") === "true") {
-            themeSwitcher.setAttribute("aria-checked", "false");
-            document.firstElementChild.classList.add('sakai-dark-theme');
-            localStorage.setItem('sakai-theme', 'dark');
+    function isPortalThemeUserPrefUnset() {
+        if (portal.userTheme === "") {
+            return true;
+        } else {
+            return false;
```

Comment:
const isPortalThemeUserPrefUnset = () => !portal.userTheme;

Reply:
This doesn't seem as readable to me, so I'm going to skip.